iocage/lib/Jail.py

```diff
@@ -0,0 +1,575 @@
+            jail=self
+        )
+
+        command = [
+            "/sbin/route",
+            "add"
+        ] + (["-6"] if (ipv6 is True) else []) + [
+            "default",
+            gateway
+        ]
+
+        self.exec(command)
+
+    def require_jail_not_existing(self):
+        if self.exists:
+            msg = f"Jail {self.humanreadable_name} already exists"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_existing(self):
+        if not self.exists:
+            msg = f"Jail {self.humanreadable_name} does not exist"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_stopped(self):
+        if self.running:
+            msg = f"Jail {self.humanreadable_name} is already running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_running(self):
+        if not self.running:
+            msg = f"Jail {self.humanreadable_name} is not running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def update_jail_state(self):
+        try:
+            stdout = subprocess.check_output([
+                "/usr/sbin/jls",
+                "-j",
+                self.identifier,
+                "-v",
+                "-h"
+            ], shell=False, stderr=subprocess.DEVNULL)
+            output = stdout.decode("utf-8").strip()
+
+            keys, values = [x.split(" ") for x in output.split("\n")]
+            self.jail_state = dict(zip(keys, values))
+
+        except:
+            self.jail_state = None
+
+    def _teardown_mounts(self):
+
+        mountpoints = list(map(
+            lambda mountpoint: f"{self.path}/root{mountpoint}",
+            [
+                "/dev/fd",
+                "/dev",
+                "/proc",
+                "/root/compat/linux/proc"
+            ]
+        ))
+
+        mountpoints += list(map(lambda x: x["destination"],
+                                list(self.config.fstab)))
+
+        for mountpoint in mountpoints:
+            if os.path.isdir(mountpoint):
+                iocage.lib.helpers.umount(
+                    mountpoint,
+                    force=True,
+                    ignore_error=True  # maybe it was not mounted
+                )
+
+    def _resolve_name(self, text):
+        jails_dataset = self.host.datasets.jails
+        for dataset in list(jails_dataset.children):
+            dataset_name = dataset.name[(len(jails_dataset.name) + 1):]
+            if text in dataset_name:
+                self.logger.debug(f"Resolved {text} to uuid {dataset_name}")
+                return dataset_name
+
+        raise Exception(f"No jail matching {text} was found")
+
+    def _get_name(self):
+        return self.config.id
+
+    def _get_humanreadable_name(self):
+
+        try:
+            uuid.UUID(self.name)
+            return str(self.name)[:8]
+        except:
```

Comment:
Only (TypeError, ValueError) here. If the name is a proper uuid it has more than 8 characters. But name could be None, so that uuid raises a TypeError)